Exploration robot module

atlantiscze opened the pull request:
- Removes mining and research modules
- Adds Exploration module. Combines features of both of these.
- Adds basic engineering capabilities to the exploration module - should be capable of basic construction necessary to, for example, set up power at an away site or perform rudimentary repairs. Metal and glass has only minimal capacities, and it lacks the specialization of engineering module.
- Adds xenoarch equipment to the module
- Adds "laser pickaxe" which is de facto a xenoarch pick which's excavation depth can be changed to values that are used by other picks. Currently only in the module's loadout. It already has quite a lot of items in its loadout, and adding ten picks would just bloat it unnecessarily.
- Intention behind these changes is to give more things to do to the module. With recent dawn of expeditions, it is not uncommon to see rounds where all focus is on exploring wrecks, and no mining operation is launched. Majority of the time the module is unusable, and lacks things to do. I opened a discussion thread on the forum, and i didn't receive any negative opinions.

:cl:
rscadd: Added exploration robot module, which replaces the mining and research modules. This module has basic engineering capabilities and combined abilities of mining and research modules.
/:cl:

atlantiscze committed:
Exploration robot module
- Removes mining and research modules
- Adds Exploration module. Combines features of both of these.
- Adds basic engineering capabilities to the exploration module - should be capable of basic construction necessary to, for example, set up power at an away site or perform rudimentary repairs. Metal and glass has only minimal capacities, and it lacks the specialization of engineering module.
- Adds xenoarch equipment to the module
- Adds "laser pickaxe" which is de facto a xenoarch pick which's excavation depth can be changed to values that are used by other picks. Currently only in the module's loadout. It already has quite a lot of items in its loadout, and adding ten picks would just bloat it unnecessarily.
- Intention behind these changes is to give more things to do to the module. With recent dawn of expeditions, it is not uncommon to see rounds where all focus is on exploring wrecks, and no mining operation is launched. Majority of the time the module is unusable, and lacks things to do. I opened a discussion thread on the forum, and i didn't receive any negative opinions.

Rebruiza commented on code/_global_vars/lists/flavor.dm:
```diff
@@ -1,8 +1,7 @@
 	"Standard", "Engineering", "Surgeon",  "Crisis",
-	"Miner",    "Janitor",     "Service",  "Clerical", "Security",
-	"Research"
+	"Exploration",    "Janitor",     "Service",  "Clerical", "Security",
```
Why does the amount of spacing vary?

Rebruiza commented on code/modules/xenoarcheaology/tools/tools_pickaxe.dm:
```diff
@@ -95,6 +95,28 @@
 
+// One pick to rule them all - configurable digging depth. Originally intended as a robot module.s
+/obj/item/weapon/pickaxe/laser
+	name = "laser pickaxe"
```
The name implies that the sprite would feature a futuristic laser look, albeit it does not. Perhaps name it "extendable pickaxe" or "adjustable pickaxe" instead?

Here is an example sprite. Feel free to use it, or for inspiration.
![image](https://user-images.githubusercontent.com/31956636/34346210-bbb5526e-e9f5-11e7-811e-da96892cc732.png)

![image](https://user-images.githubusercontent.com/31956636/34346214-c6e150d4-e9f5-11e7-86f6-aebe8dd99cb2.png)

(Should not matter much either way considering it's a borg item, but perhaps it would be nice in the future in case someone decides to implement it ingame.)

atlantiscze replied:
Thank you for the sprite. Will add.

Rebruiza commented on code/modules/xenoarcheaology/tools/tools_pickaxe.dm:
```diff
@@ -95,6 +95,28 @@
+
+/obj/item/weapon/pickaxe/laser/examine(var/mob/user)
+	. = ..()
+	to_chat(user, "It is set to excavate [excavation_amount]cm")
```
Spacing, and punctuation.

```"It is set to excavate [excavation_amount] cm."```

Rebruiza commented on code/modules/xenoarcheaology/tools/tools_pickaxe.dm:
```diff
@@ -95,6 +95,28 @@
+	to_chat(user, "It is set to excavate [excavation_amount]cm")
+
+/obj/item/weapon/pickaxe/laser/attack_self(var/mob/user)
+	var/new_setting = input("Please select excavation depth") as null|anything in list(1, 2, 4, 6, 8, 10, 12, 30)
```
```"Please select excavation depth"```

Punctuation. (Is it a title or a subtitle?)

Rebruiza commented on code/modules/xenoarcheaology/tools/tools_pickaxe.dm:
```diff
@@ -95,6 +95,28 @@
+	icon = 'icons/obj/xenoarchaeology.dmi'
+	icon_state = "pick_hand"
+	item_state = "syringe_0"
+	digspeed = 20
```
```digspeed``` varies between 20, and 30 depending if it is a smaller pickaxe, respectively a hand pickaxe. Should be negligable, irregardless.

Hand Pickaxe: https://github.com/Baystation12/Baystation12/blob/dev/code/modules/xenoarcheaology/tools/tools_pickaxe.dm#L79

12cm Pickaxe: https://github.com/Baystation12/Baystation12/blob/dev/code/modules/xenoarcheaology/tools/tools_pickaxe.dm#L91

Rebruiza commented on code/modules/xenoarcheaology/tools/tools_pickaxe.dm:
```diff
@@ -95,6 +95,28 @@
 	w_class = ITEM_SIZE_SMALL
 
+// One pick to rule them all - configurable digging depth. Originally intended as a robot module.s
+/obj/item/weapon/pickaxe/laser
```
Should this new pickaxe not go with all the other pickaxes of the same class?

See: https://github.com/Baystation12/Baystation12/blob/dev/code/modules/xenoarcheaology/tools/tools_pickaxe.dm

atlantiscze replied:
If you are referring to w_class, it is NORMAL instead of SMALL as a kind of balancing factor if it ever gets used for non-borgs (borgs don't care about w_class)

Rebruiza commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -592,47 +594,52 @@ var/global/list/robot_modules = list(
+	src.modules += new /obj/item/device/ano_scanner(src)
+	src.modules += new /obj/item/device/depth_scanner(src)
+	src.emag = new /obj/item/weapon/pickaxe/plasmacutter(src)
+
```
Stray Newline.

Rebruiza commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -592,47 +594,52 @@ var/global/list/robot_modules = list(
 	src.modules += new /obj/item/weapon/wirecutters(src)
-	src.modules += new /obj/item/weapon/crowbar(src)
+	src.modules += new /obj/item/weapon/gripper(src)
 	src.modules += new /obj/item/weapon/scalpel/laser3(src)
```
Does the exploration module really need a laser scalpel and a circular saw? I can understand the fluff behind "cutting up aliens", but they are not really made for surgery, nor do simplemobs allow for surgery.

atlantiscze replied:
Research module had these for slimes. Can remove if people don't want it to have them.

MistakeNot4892 commented:
I would recommend nuking the portable destructive analyzer while you're doing this given that it doesn't really fit a purpose on a surveyor robot and only really sees use as a means for traitor borgs to remove ion weapons from the armory.

Techhead0 commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -592,47 +594,52 @@ var/global/list/robot_modules = list(
 	src.modules += new /obj/item/weapon/gripper/miner(src)
```
It doesn't need 5 different grippers. Take `/obj/item/weapon/gripper/miner` off since it only holds a subset of `/obj/item/weapon/gripper/research`.

Techhead0 commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -592,47 +594,52 @@ var/global/list/robot_modules = list(
+	src.modules += new /obj/item/weapon/inflatable_dispenser/robot(src)
+	src.modules += new /obj/item/weapon/pickaxe/laser(src)
+	src.modules += new /obj/item/device/measuring_tape(src)
+	src.modules += new /obj/item/device/gps(src)
```
GPS would probably be better as a general verb for all borgs instead of needing a module slot to use. I suspect Crisis borgs would find it handy.

Techhead0 reviewed (commented):
In addition to the changes noted below, I don't know if it needs to also replicate engineering functionality. A combination of two old borgs PLUS stuff from another borg makes it *very* jack-of-all-trades. (And at current, it has *thirty-three* modules.)

atlantiscze committed:
Tweaks as per feedback, travis correction, minor generalization
- It is now possible to easily add items to all robot modules at once, without needing to copy-paste it everywhere
- Demonstrated using the flash (all robots already had it), crowbar (addition, allows cyborgs to move around in powerless areas) and fire extinguisher (allows for at least some maneuvering in EVA, most important modules already had it anyway).

atlantiscze commented:
Travis error should be fixed, tweaked most requested things, trimmed some less important research tools from the exploration module, added a way to add a tool to all modules at once (so far used by flash, crowbar and fire extinguisher)

Techhead0 commented:
Code mostly looks good (still have opinions on the GPS, but that's minor) but I'm gonna stew a bit on the Explorer-bot having such Engineering capabilities.

atlantiscze commented:
I tried to limit the engineering capabilities in terms of how much it can do. It still lacks specialised tools for complex engineering tasks (atmos, ..), and its material storages are only just about enough for simplest construction tasks (tear down or set up a wall, lay few tiles of wire to patch a solar array, etc.).

I agree that the GPS would be a nice addition, though i don't feel like adding it to this particular PR as my development capacity is limited right now. I might do it at some point in the future.

Techhead0 reviewed (dismissed):
I want to approve this, but it just has *so much stuff* that it's a standout module. It mines. It does xenoarch. It can build robots and work toxins. With the engineering gear on top, it's practically two-and-a-half roles in one. 

Research module already had criticism directed at it. I'm aware that many of those criticisms were directed at the portable DA. But compared to the research module, this exploration module gains much more than it's lost.

atlantiscze commented:
@Techhead0 If you wish to block the PR with changes request, please be specific on which tools you believe should be removed and why.

Also my commentary:
Xenoarch - it isn't capable of fully doing it - it is only able to excavate the anomalies properly, which is the "away" part of xenoarch/anomalies.
Building robots - I don't think this is a bad thing honestly, though i guess this functionality could be transferred over to engineering module now that robotics is under engineering and not research.
Engineering gear - Limited where possible - welding tool has only few uses before it runs out of fuel, materials have just few sheets and more complex forms (rods/rglass/etc.) are not available at all.

Techhead0 commented:
From IRC last night, since I didn't reach you with it:
> Techhead: How does this sound: Remove the engy gripper and copy anything that the Exploratory module *really* needs to the research one, leave everything else?
Techhead: Optionally, also put the 'basic engy' stuff on the standard module since it's sad.

atlantiscze commented:
I can remove the gripper, though i am not sure how you meant the "copy anything to the research one" part, as the research module was removed.

Techhead0 commented:
By "the research one" I meant `/obj/item/weapon/gripper/research` AKA "scientific gripper".

atlantiscze commented:
Ah, right, i misinterpreted that a bit. Will tweak in a bit.

atlantiscze committed:
Remove redundant gripper from the module

atlantiscze commented:
All done, sorry for the delay.

PurpleMartinJCK commented:
There's a lot of stuff this doesn't need. Welder, inflatables, all those things that make it essentially an engiborg+. Robot analyser, sheet grabber, what is all this stuff even for?

atlantiscze commented:
Sheet grabber and robot analyser are remains from research module.

Engineering tools are given to them because it is pretty safe to assume that they may have to do some engineering work (breaking into depowered derelict ships, or performing simple repairs to them). These engineering tools were balanced to be towards that "simple repairs/tasks" path by giving them very limited usage count before resupply is required.

Techhead0 commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -559,93 +551,91 @@ var/global/list/robot_modules = list(
 	subsystems = list(/datum/nano_module/supply)
-	channels = list("Supply" = 1, "Science" = 1)
+	channels = list("Supply" = 1, "Science" = 1, "Exploration" = 1)
 	networks = list(NETWORK_MINE)
```
Missed this, should have NETWORK_EXPLO now.

Techhead0 replied:
Hey.

atlantiscze replied:
Corrected

Sunbeamstress commented:
@atlantiscze Hey where are you at on this? I'm really looking forward to this change!

atlantiscze commented:
Sorry, but i'll need few more days before i can work on this again. There were some high priority IRL things which i've been spending all my time handling recently.

Coderofthecode commented:
Keep the destructive analyser, please. It's the only reliable way to actually do RnD-science as a research borg, as your science gripper can't pick up the majority of the protolathe list. In addition, removing the saw and scalpel and such prevents them from doing xenobiology whatsoever, the wrench/screwdriver/crowbar doing any robotics work despite being geared for it and they can't deal with many of the tools to do anomalies as is, ignoring the frequent EMPs involved. This effectively means the borg can't do any research at all and realistically only mine. 

Which... um? I recommend buffing the miner into the explorer and keeping the research borg as is if you're going to do this, because as is this basically just deletes a cyborg for research.

atlantiscze committed:
Merge branch 'dev' into 2017-12-26-ExplorationModule

atlantiscze commented:
Made an attempt to fix this using online editor. _What could possibly go wrong.._

Rebruiza commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -559,93 +551,92 @@ var/global/list/robot_modules = list(
+	src.modules += new /obj/item/device/gps(src)
+	src.modules += new /obj/item/device/ano_scanner(src)
+	src.modules += new /obj/item/device/depth_scanner(src)
+	src.emag = new /obj/item/weapon/pickaxe/plasmacutter(src)
```
/obj/item/weapon/gun/energy/plasmacutter

The pathing changed due to the plasmacutter being made into a ranged weapon.

ghost commented:
See comment above, @atlantiscze.

PurpleMartinJCK commented:
As long as you continue work on this it has been updated recently enough it's not at risk of being closed.

atlantiscze committed:
Type correction

atlantiscze commented:
Should be resolved.

atlantiscze committed:
Network correction

babydoll commented on code/modules/mob/living/silicon/robot/robot_modules.dm:
```diff
@@ -559,93 +551,92 @@ var/global/list/robot_modules = list(
-	channels = list("Supply" = 1, "Science" = 1)
-	networks = list(NETWORK_MINE)
+	channels = list("Supply" = 1, "Science" = 1, "Exploration" = 1)
+	networks = list(NETWORK_EXPLO)
```
`NETWORK_EXPLO` is Torch exclusive

babydoll commented:
@atlantiscze

comma commented:
I think it's bit beyond stale stage.
